### ISSUE TITLE
twisted.trial.unittest.TestCase.addCleanup causes cleanups to run more than once

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10320]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#10320|
|Type|enhancement|
|Created|2022-03-27 16:26:01Z|

```
from twisted.trial.unittest import TestCase

class Tests(TestCase):
    def test_foo(self):
        self.addCleanup(bar)

```

If run with `trial -u thosetests` then `bar` will be called once after the first iteration, twice after the second iteration, etc.

Compare this to the behavior of `SynchronousTestCase.addCleanup` where cleanup functions are discarded when they are run - so a test like the above would cause `bar` to be run once after the first iteration, once after the second iteration, etc.

<details><summary>Searchable metadata</summary>

```
trac-id__10320 10320
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__None None
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__None None
time__1648398361324227 1648398361324227
changetime__1648429370405359 1648429370405359
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

https://github.com/twisted/twisted/pull/1708

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph set owner to @exarkun|
|-|-|

Reviewed on github.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set status to `closed`|
|-|-|

Merged in 99a9c0cf4242619387bef57d510a59b1d87e9b51
